frontend/src/routes/(authenticated)/project/[project_code]/BulkAddProjectMembers.svelte

```diff
@@ -0,0 +1,79 @@
+<script lang="ts">
+  import { BadgeButton } from '$lib/components/Badges';
+  import { DialogResponse, FormModal } from '$lib/components/modals';
+  import { Input, ProjectRoleSelect, TextArea, passwordFormRules } from '$lib/forms';
+  import { ProjectRole } from '$lib/gql/types';
+  import t from '$lib/i18n';
+  import { z } from 'zod';
+  import { _bulkAddProjectMembers } from './+page';
+  import { useNotifications } from '$lib/notify';
+  import { hash } from '$lib/util/hash';
+  import { AdminContent } from '$lib/layout';
+
+  export let projectId: string;
+  const schema = z.object({
+    usernamesText: z.string().min(1, $t('register.name_missing')),
+    password: passwordFormRules($t),
+    role: z.enum([ProjectRole.Editor, ProjectRole.Manager]).default(ProjectRole.Editor),
+  });
+
+  let formModal: FormModal<typeof schema>;
+  $: form = formModal?.form();
+
+  let createdCount: number | undefined = undefined;
+  // let usernameConflicts: string[] = [];
+
+  const { notifySuccess } = useNotifications();
+
+  async function openModal(): Promise<void> {
+    const { response, formState } = await formModal.open(async () => {
+      const passwordHash = await hash($form.password);
+      const usernames = $form.usernamesText.split('\n').map(s => s.trim());
+      const { error, data } = await _bulkAddProjectMembers({
+        projectId,
+        passwordHash,
+        usernames,
+        role: $form.role,
+      });
+
+      // TODO: Handle this case
+      // if (error?.byType('NotFoundError')) {
+      //   return { email: [$t('project_page.bulk_add_members.project_not_found')] };
+      // }
+
+      createdCount = data?.bulkAddProjectMembers.bulkAddProjectMembersResult?.createdCount;
+      // usernameConflicts = data?.bulkAddProjectMembers.bulkAddProjectMembersResult?.usernameConflicts ?? [];
+      return error?.message;
+    });
+    if (response === DialogResponse.Submit) {
+      // const message = userInvited ? 'member_invited' : 'add_member';
+      notifySuccess($t(`project_page.notifications.bulk_add_members`, { count: createdCount ?? 0 }));
+      // TODO: Display username conflicts somewhere as well
+    }
+  }
+</script>
+
+<AdminContent>
+  <BadgeButton type="badge-success" icon="i-mdi-account-plus-outline" on:click={openModal}>
+    {$t('project_page.bulk_add_members.add_button')}
+  </BadgeButton>
+
+  <FormModal bind:this={formModal} {schema} let:errors>
+    <span slot="title">{$t('project_page.bulk_add_members.modal_title')}</span>
+    <Input
+      id="password"
+      type="password"
+      label={$t('project_page.bulk_add_members.shared_password')}
+      bind:value={$form.password}
+      error={errors.password}
+    />
+    <ProjectRoleSelect bind:value={$form.role} error={errors.role} />
```

Comment:
I'm not sure if this has already been discussed, but:
We're planning on not letting these users login to the UI, right?
So, I think we should always give them the `Editor` role instead of letting the admin choose.

Also, `AddProjectMember`/`ChangeProjectMemberRole` should prevent users that can't login from being made managers.

Reply:
My recollection of our discussions was that we would let them log in to the UI, and if they add an email address (and verify it), we would grant them more privileges, such as being able to be managers.

My first design had them being forced into the Editor role, and I decided that no, admins shouldn't be restricted too much. But you're right, we don't want any project managers with shared passwords and no email addresses. I'll lock this down to Editor and remove the dropdown. I'll also make that change in AddProjectMember et al.

Reply:
Carried on the conversation here: https://github.com/sillsdev/languageforge-lexbox/pull/621#discussion_r1514704757
...for some reason